Add Open Energy Modeling at JuMP-dev 2024 (#156)
jump-dev/jump-dev.github.io: _posts/2024-09-04-open-energy-models.md

is a summary of my notes from watching the energy-related talks.

### Contents

 * [Applied optimization with JuMP at SINTEF](#applied-optimization-with-jump-at-sintef)
 * [Introduction to TulipaEnergyModel.jl](#introduction-to-tulipaenergymodeljl)
 * [SpineOpt.jl: A highly adaptable modelling framework for multi-energy systems](#spineoptjl-a-highly-adaptable-modelling-framework-for-multi-energy-systems)
 * [Solving the Market-to-Market Problem in Large Scale Power Systems](#solving-the-market-to-market-problem-in-large-scale-power-systems)
 * [PiecewiseAffineApprox.jl](#piecewiseaffineapproxjl)

## Applied optimization with JuMP at SINTEF

_Speaker: Truls Flatberg @trulsf_

<iframe width="560" height="315" src="https://www.youtube.com/embed/-a9-ToFiT8E?si=rpQeHt1__-lwxjXT" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this [prize winning](prize/jump-dev-2024) talk, Truls discussed how they have
been using JuMP to build and solve large scale optimization models at SINTEF.

> SINTEF has been an active attendees of JuMP-dev recently, speaking about

Author: ```suggestion
SINTEF have been active attendees of JuMP-dev recently, speaking about
```

 * [Solving the Market-to-Market Problem in Large Scale Power Systems](#solving-the-market-to-market-problem-in-large-scale-power-systems)
 * [PiecewiseAffineApprox.jl](#piecewiseaffineapproxjl)

## Applied optimization with JuMP at SINTEF

_Speaker: Truls Flatberg @trulsf_

<iframe width="560" height="315" src="https://www.youtube.com/embed/-a9-ToFiT8E?si=rpQeHt1__-lwxjXT" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this [prize winning](prize/jump-dev-2024) talk, Truls discussed how they have
been using JuMP to build and solve large scale optimization models at SINTEF.

SINTEF have been active attendees of JuMP-dev recently, speaking about
[UnitJuMP.jl](https://www.youtube.com/watch?v=JQ6_LZfYRqg) and
[SparseVariables.jl](https://www.youtube.com/watch?v=YuDvfZo9W5A) at
[JuMP-dev 2022](/meetings/juliacon2022/), and
[TimeStruct.jl](https://www.youtube.com/watch?v=Hz6AL5kClKU) and

> [energy system models](https://www.youtube.com/watch?v=fARbeM8sANA) at

Author: ```suggestion
[EnergyModelsX](https://www.youtube.com/watch?v=fARbeM8sANA) at
```

the simulation models.

In regard to feature requests for future versions of JuMP, Jose again referenced
the need for efficient re-solves of optimization models with parameters. We
have, over the years, experimented with a number of different approaches in
Sienna, including the now largely defunct [ParameterJuMP.jl](https://github.com/JuliaStochOpt/ParameterJuMP.jl),
the intended replacement [ParametricOptInterface.jl](https://github.com/jump-dev/ParametricOptInterface.jl),
and other work-arounds like adding new `VariableIndex in EqualTo` constraints.
Ultimately, each way has a different trade-off (and it also depends on the size
of the model that you are trying to solve), and we haven't found something that
we are completely happy with.

My main takeaways are that we should include benchmarks that use
ParametricOptInterface.jl in our [jump-dev/open-energy-modeling-benchmarks](https://github.com/jump-dev/open-energy-modeling-benchmarks)
repository, and that JuMP's ability to efficiently solve a sequence of similar
problems with some changes in the problem parameters is both a main selling
point of using JuMP over alternative tools and a current bottleneck in existing

> workflows.

Reviewer: Sienna is also one of the few tools that has a workflow with multiple resolves as part of the features and needs since it is the only JuMP application focused on simulations.

Author: I don't know if this is true. SpineOpt has a rolling-horizon feature, and so does @sstromer's model.

Reviewer 2: The same is currently in development for EnergyModelsX.

(It's notable to see that the production quality of JuMP-dev videos is much
improved over the years.) The JuMP documentation also contains a tutorial,
[Approximating nonlinear functions](https://jump.dev/JuMP.jl/stable/tutorials/linear/piecewise_linear/),
which explains how to construct the linearizations manually.

The fact that multiple groups have now developed similar packages suggests that
there is a need for us to integrate these packages closer into JuMP. As an
immediate improvement, I have added `PiecewiseAffineApprox.jl` to the script we
use to check that new releases of JuMP do not break downstream packages in
[JuMP#3817](https://github.com/jump-dev/JuMP.jl/pull/3817). I have also opened
an issue,
[PiecewiseLinearOpt.jl#49](https://github.com/joehuchette/PiecewiseLinearOpt.jl/issues/49),
to discuss moving that package to `jump-dev`. In the medium term, it might be
helpful to combine the two packages into a single utility package that has a
broad range of functionality for constructing piecewise linear approximations.

[_Back to contents_](#contents)

## [2023] How JuMP Enables Abstract Energy System Models

> _Speaker: Stefan Strömer @sstromer_

Author: ```suggestion
_Speaker: Stefan Strömer @sstroemer_
```

One useful part of Diego's talk was how he linked to various issues in the
SpineOpt.jl repository. If I had to summarize the common root cause of the
issues, it is that managing the input and output of a large quantity of JuMP
models is hard! In Spine's case, a frequent use-case is solving a year of power
system operational problems at hourly resolution via a rolling horizon model of
48 hour blocks rolling forward 24 hours at each step.

Diego's experience reinforces for me the need to include rolling horizon
simulations in our set of benchmark instances that we are building in the
[jump-dev/open-energy-modeling-benchmarks](https://github.com/jump-dev/open-energy-modeling-benchmarks)
repository.

[_Back to contents_](#contents)

## [2024] Solving the Market-to-Market Problem in Large Scale Power Systems

_Speaker: Jose Daniel Lara @jd-lara_

<iframe width="560" height="315" src="https://www.youtube.com/embed/N-jDHickaTc?si=T7eQnTopbCCtES3X" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

> In this talk Jose gave an overview and spoke about recent updates to

Reviewer: ```suggestion
In this talk Jose Daniel gave an overview and spoke about recent updates to
```

[Sienna](https://www.nrel.gov/analysis/sienna.html) (nee PowerSimulations.jl).
The talk follows his previous talks at [JuMP-dev 2019](#2019-powersimulationsjl)
and [JuMP-dev 2023](https://www.youtube.com/watch?v=J7VbCKsnTvQ).

A key feature of Sienna is that it is designed for problems with _O(10⁸)_
variables and constraints. They are building and solving simulation models of
the Eastern Interconnection, which is one of, it not the, largest power system
in the world (it has 150,000 buses and 270,000 lines).

> A key lesson that I took from Jose's talk is that they have already encountered

Reviewer: ```suggestion
A key lesson that I took from Jose Daniel's talk is that they have already encountered
```

(if you watch the video, the package is now registered!) which is a JuMP
extension for working with time-structured models that often arise in planning
problems.

Truls points out that a key feature of JuMP is its modularity, and how they can
invest in the develop of small utility packages like TimeStruct.jl, and then
re-use the utility package across many different application models.

If you are building a JuMP model with time as a significant component, we
recommend that you try out
[TimeStruct.jl](https://github.com/sintefore/TimeStruct.jl).

[_Back to contents_](#contents)

## [2023] Designing a Flexible Energy System Model Using Multiple Dispatch

_Speaker: Julian Straus @JulStraus_

<iframe width="560" height="315" src="https://www.youtube.com/embed/fARbeM8sANA?si=tETN7GSZNzW5ZzCU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk, Julian discusses [EnergyModelsX](https://github.com/EnergyModelsX)
that he is building at SINTEF.

> I found it useful to compare the talk to Jose's [[2019] PowerSimulations.jl](#2019-powersimulationsjl).

Reviewer: ```suggestion
I found it useful to compare the talk to Jose Daniel's [[2019] PowerSimulations.jl](#2019-powersimulationsjl).
```

If multiple groups arrive at the same set of ideas, I think it demonstrates that

> hierarchical models that leverage Julia's multiple dispatch is the "right" way

Reviewer: We could reference from JuMP-dev 2024 that Pyomo arrived to a similar design of implementing a multiple dispatch like capability

Author: I think this is slightly different. Pyomo would say build models using blocks etc. They're using multiple dispatch internally to simplify the expression tree walkers.

_Speaker: Jose Daniel Lara @jd-lara_

<iframe width="560" height="315" src="https://www.youtube.com/embed/JAHjZYiIJeI?si=euyA8WzK-lEs_Pg9" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

This talk is the first talk by Jose at a JuMP-dev on PowerSimulations.jl (now
called Sienna). I found it useful to compare this talk to
[his talk at JuMP-dev 2024](#2024-solving-the-market-to-market-problem-in-large-scale-power-systems).

Jose described how the three design principles of PowerSimulations were
flexibility, modularity, and scalability, and how all three of these were
achieved by using Julia's multiple dispatch to build different mathematical
formulations based on the input data. I found it interesting that Jose hits upon
on many of the same themes that Julian did in his 2023 talk
[[2023] Designing a Flexible Energy System Model Using Multiple Dispatch](#2023-designing-a-flexible-energy-system-model-using-multiple-dispatch).
This suggests to me that we can better share lessons learned between developers.

> Jose also talks about the issue with parameters and time-series data. Five years

Reviewer: ```suggestion
Jose Daniel also talks about the issue with parameters and time-series data. Five years
```
